File: Python/linkedLists.py
class Node:
    def __init__(self, data):
        """
        Initializes a Node element with value data and empty next and prev.

        Args:
            data: any type of data you want in the node. 
        """

        self.value = data
        self.next = None
        self.prev = None

class LinkedList:
    def __init__(self):
        """
        Initialize an empty LinkedList object with no set head or tail and a length of 0.
        """

        self.head = None
        self.tail = None
        self.length = 0

    # 
    # 
    # INSERT NODES
    # 
    # 
    def insertFirst(self, data) -> None:
        """
        Inserts Node at the front of the list.

        Args:
            data: any type of data you want in the node.
        """

        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

        self.length += 1

    def insertLast(self, data) -> None:
        """
        Inserts Node at the back of the list.

        Args:
            data: any type of data you want in the node.
        """

        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

        self.length += 1

    def insertMiddle(self, index, data) -> None:
        """
        Inserts Node at the stated index in the list.

        Args:
            index: the position of the new Node, 0 is the first position.
            data: any type of data you want in the node.
        Raises:
            IndexError: If the stated index is negative or bigger than the length of the list.
        """

        new_node = Node(data)

        if self.head is None and index > 0:
            raise IndexError("List is empty, index out of bounds")
        if (index > self.length and self.length > 0) or index < 0:
            raise IndexError("Index out of bounds")
        if index == 0:
            self.insertFirst(data)
        if index == self.length:
            self.insertLast(data)

        count = 0
        prev_node = self.head

        while index-1 > count:
            prev_node = prev_node.next
            count += 1

        next_node = prev_node.next

        new_node.prev = prev_node
        new_node.next = next_node

        prev_node.next = new_node
        next_node.prev = new_node

        self.length += 1

    # 
    # 
    # REMOVE NODES
    # 
    # 
    def removeFirst(self) -> None:
        """
        Removes the node at the front of the list.
        """

        if self.head is None:
            raise IndexError("List is empty")
        
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            deleted_node = self.head
            next_node = deleted_node.next

            deleted_node.next = None
            next_node.prev = None

            self.head = next_node

        self.length -= 1
        
    
    def removeLast(self) -> None:
        """
        Removes the node at the back of the list.
        """

        if self.head is None:
            raise IndexError("List is empty")
        
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            deleted_node = self.tail
            prev_node = deleted_node.prev

            deleted_node.prev = None
            prev_node.next = None

            self.tail = prev_node

        self.length -= 1
    
    def removeFromIndex(self, index) -> None:
        """
        Removes the node at stated index of the list.

        Args:
            index: the index of the deleted Node, 0 is the first position.
        """

        if self.head is None and index > 0:
            raise IndexError("List is empty, index out of bounds")
        if (index > self.length and self.length > 0) or index < 0:
            raise IndexError("Index out of bounds")
        if index == 0:
            self.removeFirst()
        if index == self.length-1:
            self.removeLast()

        prev_node = self.head

        count = 0

        while index-1 > count:
            prev_node = prev_node.next
            count += 1

        deleted_node = prev_node.next
        next_node = deleted_node.next

        deleted_node.prev = None
        deleted_node.next = None

        next_node.prev = prev_node
        prev_node.next = next_node

        self.length -= 1

    # 
    # 
    # GET NODE VALUES
    # 
    # 
    def getFirst(self) -> any:
        """
        Returns the head value of the list.

        Return:
            Variable: value of the head Node.
        """

        if self.head is None:
            raise IndexError("List is empty")
        
        return self.head.value
    
    def getLast(self) -> any:
        """
        Returns the tail value of the list.

        Returns:
            Variable: value of the tail Node.
        """

        if self.head is None:
            raise IndexError("List is empty")
        
        return self.tail.value
    
    def getValueFromIndex(self, index) -> any:
        """
        Returns the value of the Node from the stated index.

        Args:
            index: the index of the Node you want to get the value from, list starts at 0.

        Returns:
            Variable: value of the Node at the stated index.
        """

        if self.head is None and index > 0:
            raise IndexError("List is empty, index out of bounds")
        if (index > self.length and self.length > 0) or index < 0:
            raise IndexError("Index out of bounds")
        if index == 0:
            return self.getFirst()
        if index == self.length-1:
            return self.getLast()
        
        current_node = self.head

        count = 0

        while index > count:
            current_node = current_node.next
            count += 1

        return current_node.value
    
    def getLength(self) -> int:
        """
        Returns the length of the list

        Returns:
            int: the length of the LinkedList
        """
        return self.length
    
    def __str__(self) -> str:
        """
        Returns an array of the values in the list

        Returns:
            str: a string looking like an array, not actual array.
        """

        if self.head == None:
            raise IndexError("List is empty")

        out_string = "["
        out_string += str(self.head.value)

        current_node = self.head

        while current_node.next is not None:
            current_node = current_node.next
            out_string += ", "
            out_string += str(current_node.value)

        out_string += "]"

        return out_string


if __name__ == "__main__":
    linked_list = LinkedList()
    linked_list.insertFirst(1)
    linked_list.insertFirst(2)
    linked_list.insertLast(9)
    print(linked_list.getFirst())
    print(linked_list)

File: Java/LinkedLists.java
package LinkedLists.Java;
import java.util.ArrayList;
import java.util.List;

public class LinkedLists<T> {
    Node<T> head;
    Node<T> tail;
    int len;

    public LinkedLists() {
        head = null;
        tail = null;
        len = 0;
    }

    //
    //
    // INSERT NODES
    //
    //
    void insertFirst(T data) {
        Node<T> new_node;
        Node<T> current_node;

        new_node = new Node<T>(data);
        if (this.head == null) {
            this.head = new_node;
            this.tail = new_node;
        } else {
            current_node = this.head;
            new_node.next = current_node;
            current_node.prev = new_node;
            this.head = new_node;
        }

        this.len++;
    }

    void insertLast(T data) {
        Node<T> new_node;
        Node<T> current_node;

        new_node = new Node<T>(data);
        if (this.head == null) {
            this.head = new_node;
            this.tail = new_node;
        } else {
            current_node = this.tail;
            new_node.prev = current_node;
            current_node.next = new_node;
            this.tail = new_node;
        }

        this.len++;
    }

    void insertMiddle(int index, T data) {
        Node<T> new_node;
        Node<T> prev_node;
        Node<T> next_node;

        if (this.head == null && index > 0) {
            throw new NullPointerException("List is empty, so index is out of bounds!");
        }
        if ((index > this.len && this.len > 1) || index < 0) {
            throw new ArrayIndexOutOfBoundsException("Index is out of bounds!");
        }
        if (index == 0) {
            insertFirst(data);
            return;
        }
        if (index == this.len) {
            insertLast(data);
            return;
        }
        
        new_node = new Node<T>(data);

        int count = 0;

        prev_node = this.head;
        while (index-1 > count) {
            prev_node = prev_node.next;
            count++;
        }

        next_node = prev_node.next;

        // Sets the next and previous for the new_node
        new_node.prev = prev_node;
        new_node.next = next_node;

        // Sets the new_node as next and prev for their respective next and prev nodes
        prev_node.next = new_node;
        next_node.prev = new_node;

        this.len++;
    }

    //
    //
    // REMOVE NODES
    //
    //
    void removeFirst() {
        Node<T> deleted_node;
        Node<T> next_node;

        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }

        if (this.len == 1) {
            this.head = null;
            this.tail = null;
        } else {
            deleted_node = this.head;
            next_node = deleted_node.next;

            deleted_node.next = null;
            next_node.prev = null;

            this.head = next_node;
        }

        this.len--;
    }

    void removeLast() {
        Node<T> deleted_node;
        Node<T> prev_node;

        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }

        if (this.len == 1) {
            this.head = null;
            this.tail = null;
        } else {
            deleted_node = this.tail;
            prev_node = deleted_node.prev;

            deleted_node.prev = null;
            prev_node.next = null;

            this.tail = prev_node;
        }
        
        this.len--;
    }

    void removeFromIndex(int index) {
        Node<T> deleted_node;
        Node<T> prev_node;
        Node<T> next_node;

        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }
        if (index >= this.len || index < 0) {
            throw new ArrayIndexOutOfBoundsException("Index is out of bounds!");
        }
        if (index == 0) {
            removeFirst();
            return;
        }
        if (index == this.len-1) {
            removeLast();
            return;
        }

        prev_node = this.head;

        int count = 0;

        while (index-1 > count) {
            prev_node = prev_node.next;
            count++;
        }

        deleted_node = prev_node.next;
        next_node = deleted_node.next;

        deleted_node.prev = null;
        deleted_node.next = null;

        next_node.prev = prev_node;
        prev_node.next = next_node;

        this.len--;
    }

    //
    //
    // GET NODE VALUES
    //
    //
    T getFirst() {
        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }
        return this.head.value;
    }

    T getLast() {
        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }
        return this.tail.value;
    }

    T getValueFromIndex(int index) {
        Node<T> current_node;

        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }
        if (index >= this.len || index < 0) {
            throw new ArrayIndexOutOfBoundsException("Index is out of bounds!");
        }
        if (index == 0) {
            return getFirst();      
        }
        if (index == this.len-1) {
            return getLast();  
        }

        current_node = this.head;

        int count = 0;

        while (index > count) {
            current_node = current_node.next;
            count++;
        }

        return current_node.value;
    }

    int getLength() {
        return this.len;
    }

    // Returns the entire list as a String looking like an array
    String getStringList() {
        Node<T> current_node;
        String out_string;

        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }

        out_string = "[";
        
        current_node = this.head;
        out_string += String.valueOf(current_node.value);
        while (current_node.next != null) {
            current_node = current_node.next;
            out_string += ", ";
            out_string += String.valueOf(current_node.value);
        }

        out_string += "]";

        return out_string;
    }

    // Creates an ArrayList of the LinkedList and returns it
    List<Object> getArrayList() {
        Node<T> current_node;
        List<Object> outArray = new ArrayList<>();

        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }

        current_node = this.head;
        outArray.add(current_node.value);
        while (current_node.next != null) {
            current_node = current_node.next;
            outArray.add(current_node.value);
        }

        return outArray;
    }

    // Prints the entire list, line by line
    void printLinkedList() {
        Node<T> current_node;

        if (this.head == null) {
            throw new NullPointerException("List is empty!");
        }
        current_node = this.head;
        System.out.println(current_node.value);
        while (current_node.next != null) {
            current_node = current_node.next;
            System.out.println(current_node.value);
        }
    }
}